django/core/handlers/exception.py

```diff
@@ -98,6 +98,7 @@ def response_for_exception(request, exc):
         security_logger = logging.getLogger('django.security.%s' % exc.__class__.__name__)
         security_logger.error(
             str(exc),
+            exc_info=exc,
```

Comment:
Is there any reason to use `exc` instead of `sys.exc_info()`?, as in other cases :thinking:

Reply:
No... But looking at the implementation in `Logger._log()`:

```
        if exc_info:
            if isinstance(exc_info, BaseException):
                exc_info = (type(exc_info), exc_info, exc_info.__traceback__)
            elif not isinstance(exc_info, tuple):
                exc_info = sys.exc_info()
```

If you pass an exception instance, it will decompose it into the `(type, value, traceback)` tuple. [^1] Instead of calling `sys.exc_info()`, perhaps we should hand that off to the logger and pass `exc_info=True` everywhere.

Or, perhaps we should instead pass `exc_info=exc` everywhere because `response_for_exception()` is called within the exception block, but could possibly (if unlikely) be called in a different context where `sys.exc_info()` would give the wrong exception.

[^1]: Out of interest, there is currently a lot of change going into 3.11 that ditches this 3-tuple mess for exceptions under the hood as it is pointless. Obviously the API will have to keep supporting this for a long time, perhaps forever.

Reply:
There's a difference between this and other cases -- this calls `Logger.error()` directly, while others go through `log_response()`.

I'm not sure that, in general, taking threads and async in consideration, calling `sys.exc_info()` out of an `except:` block is valid (although it is done here in all the other cases).

However, if it is, we might as well call `security_logger.exception()` instead.

Reply:
> There's a difference between this and other cases -- this calls `Logger.error()` directly, while others go through `log_response().`

Yes, I did notice this. The main difference is that the message is logged before the response is created. That prevents use of `log_response()` by passing `response`, although you can pass `logger` and `level` independently to that function instead.

> However, if it is, we might as well call `security_logger.exception()` instead.

True. Calling `.exception(…)` is the same as calling `.error(…, exc_info=True)`.

> I'm not sure that, in general, taking threads and async in consideration, calling sys.exc_info() out of an except: block is valid (although it is done here in all the other cases).

But I guess this is the main thing: If we already have an instance of the exception, why not pass that rather than look it up again with `sys.exc_info()` and run the risk that, somehow, however unlikely, we end up with the wrong exception.

Reply:
> But I guess this is the main thing: If we already have an instance of the exception, why not pass that rather than look it up again with `sys.exc_info()` and run the risk that, somehow, however unlikely, we end up with the wrong exception.

💯